alwaysEmptyObject and alwaysEmptyArray do not always return empty array or object

Output from `alwaysEmptyArray` and `alwaysEmptyObject` cannot be trusted to be always empty. Specially if their outputs are being passed around as inputs to third party libraries. This can lead to hard to track bugs and odd behaviors.

```js
R_.alwaysEmptyArray(); // []
R_.alwaysEmptyArray().push(42); 
R_.alwaysEmptyArray(); // [42] <- return value has been changed forever for every other subsequent call
```

I propose we either: 

- _Change the name of the functions_ to reflect the above fact.
- Return a `R.clone` of the `emptyArray` or `emptyObject` each time:

```js
const alwaysEmptyArray = () => clone(emptyArray);
```

- Leverage the new (unreleased as of yet) [`R.thunkify`](https://github.com/ramda/ramda/pull/2551) function to create a point-free reusable factory of clones and implement other `always*` functions based off of it.

```js
const alwaysNew = thunkify(clone);
const alwaysEmptyArray = alwaysNew(emptyArray);
```

> DISCLAIMER: I was the author of the `thunkify` PR.

@tommmyy I was trying to get a PR together for this today and found out that I cannot install the project on a fresh clone (running on macOS, `node` `v10.4.1 and `npm` `6.1.0`).

The installation for `fsevents` seems to fail. It tries to download binaries, errors and defaults to compile from source which also fails.

```sh
❯ git clone git@github.com:nfantone/ramda-extension.git
❯ cd ramda-extension
❯ npm i

> fsevents@1.1.3 install /Users/nfantone/Development/js/ramda-extension/node_modules/fsevents
> node install

node-pre-gyp ERR! Tried to download(404): https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.1.3/fse-v1.1.3-node-v64-darwin-x64.tar.gz
node-pre-gyp ERR! Pre-built binaries not found for fsevents@1.1.3 and node@10.4.1 (node-v64 ABI, unknown) (falling back to source compile with node-gyp)
node-pre-gyp ERR! Tried to download(undefined): https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.1.3/fse-v1.1.3-node-v64-darwin-x64.tar.gz
node-pre-gyp ERR! Pre-built binaries not found for fsevents@1.1.3 and node@10.4.1 (node-v64 ABI, unknown) (falling back to source compile with node-gyp)

[...bunch more...]

make: *** [Release/obj.target/fse/fsevents.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at ChildProcess.emit (events.js:182:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12
gyp ERR! System Darwin 17.6.0

[...even more...]
```

Removing `package-lock.json` solved the issue for me. Incidentally, I found out that it _does work_ with `yarn` out of the box. I'm guessing one of the lockfiles is not up-to-date?

--- a/packages/ramda-extension/src/__tests__/alwaysEmptyArray-test.js
+++ b/packages/ramda-extension/src/__tests__/alwaysEmptyArray-test.js
@@ -4,4 +4,8 @@ describe('alwaysEmptyArray', () => {
 	it('returns []', () => {
 		expect(alwaysEmptyArray()).toEqual([]);
 	});
+
+	it('always returns a new array', () => {
+		expect(alwaysEmptyArray()).not.toBe(alwaysEmptyArray());
+	});
 });

--- a/packages/ramda-extension/src/__tests__/alwaysEmptyObject-test.js
+++ b/packages/ramda-extension/src/__tests__/alwaysEmptyObject-test.js
@@ -4,4 +4,8 @@ describe('alwaysEmptyObject', () => {
 	it('returns {}', () => {
 		expect(alwaysEmptyObject()).toEqual({});
 	});
+
+	it('always returns a new object', () => {
+		expect(alwaysEmptyObject()).not.toBe(alwaysEmptyObject());
+	});
 });

--- a/packages/ramda-extension/src/alwaysEmptyArray.js
+++ b/packages/ramda-extension/src/alwaysEmptyArray.js
@@ -1,8 +1,8 @@
-import { always } from 'ramda';
+import alwaysNew from './internal/alwaysNew';
 import { emptyArray } from './internal/primitives';
 
 /**
- * Always returns empty array.
+ * Always returns a new empty array.
  *
  * @func
  * @category Function
@@ -13,6 +13,6 @@ import { emptyArray } from './internal/primitives';
  *
  * @sig a -> Array
  */
-const alwaysEmptyArray = always(emptyArray);
+const alwaysEmptyArray = alwaysNew(emptyArray);
 
 export default alwaysEmptyArray;

--- a/packages/ramda-extension/src/alwaysEmptyObject.js
+++ b/packages/ramda-extension/src/alwaysEmptyObject.js
@@ -1,8 +1,8 @@
-import { always } from 'ramda';
+import alwaysNew from './internal/alwaysNew';
 import { emptyObject } from './internal/primitives';
 
 /**
- * Always returns empty object.
+ * Always returns a new empty object.
  *
  * @func
  * @category Function
@@ -13,6 +13,6 @@ import { emptyObject } from './internal/primitives';
  *
  * @sig a -> Object
  */
-const alwaysEmptyObject = always(emptyObject);
+const alwaysEmptyObject = alwaysNew(emptyObject);
 
 export default alwaysEmptyObject;

--- a/packages/ramda-extension/src/internal/alwaysNew.js
+++ b/packages/ramda-extension/src/internal/alwaysNew.js
@@ -8,7 +8,7 @@ import thunkify from '../internal/thunkify';
  * @func
  * @example
  *
- * 	const alwaysNewArray = alwaysNew([]);
+ *  const alwaysNewArray = alwaysNew([]);
  *  const a = alwaysNewArray();
  *  const b = alwaysNewArray();
  *  // a !== b

--- a/packages/ramda-extension/src/internal/alwaysNew.js
+++ b/packages/ramda-extension/src/internal/alwaysNew.js
@@ -1,0 +1,19 @@
+import { clone } from 'ramda';
+import thunkify from '../internal/thunkify';
+
+/**
+ * Returns a function that creates new instances of whatever argument
+ * is passed in each time it's called.
+ *
+ * @func
+ * @example
+ *
+ * 	const alwaysNewArray = alwaysNew([]);
+ *  const a = alwaysNewArray();
+ *  const b = alwaysNewArray();
+ *  // a !== b
+ *
+ */
+const alwaysNew = thunkify(clone);
+
+export default alwaysNew;

--- a/packages/ramda-extension/src/internal/thunkify.js
+++ b/packages/ramda-extension/src/internal/thunkify.js
@@ -1,0 +1,19 @@
+import { curryN } from 'ramda';
+
+/**
+ * Creates a thunk out of a function. A thunk delays a calculation until
+ * its result is needed, providing lazy evaluation of arguments.
+ *
+ * @func
+ * @example
+ *
+ *      R_.thunkify(R.identity)(42)(); //=> 42
+ *      R_.thunkify((a, b) => a + b)(25, 17)(); //=> 42
+ * @sig ((a, b, ..., j) -> k) -> (a, b, ..., j) -> (() -> k)
+ */
+const thunkify = (fn) => curryN(fn.length, (...args) => () => fn(...args));
+
+// NOTE (23/06/18): `thunkify` was added to Ramda core but it's
+// still unreleased. We should remove this module once it becomes available.
+// @see https://github.com/ramda/ramda/pull/2551
+export default thunkify;

--- a/packages/ramda-extension/src/internal/thunkify.js
+++ b/packages/ramda-extension/src/internal/thunkify.js
@@ -1,4 +1,6 @@
-import { curryN } from 'ramda';
+import { curry, apply, useWith, unapply } from 'ramda';
+
+const alwaysApply = curry((fn, args) => () => apply(fn, args));
 
 /**
  * Creates a thunk out of a function. A thunk delays a calculation until
@@ -9,9 +11,10 @@ import { curryN } from 'ramda';
  *
  *      R_.thunkify(R.identity)(42)(); //=> 42
  *      R_.thunkify((a, b) => a + b)(25, 17)(); //=> 42
+ *
  * @sig ((a, b, ..., j) -> k) -> (a, b, ..., j) -> (() -> k)
  */
-const thunkify = (fn) => curryN(fn.length, (...args) => () => fn(...args));
+const thunkify = useWith(unapply, [alwaysApply]);
 
 // NOTE (23/06/18): `thunkify` was added to Ramda core but it's
 // still unreleased. We should remove this module once it becomes available.

